server/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -219,7 +383,29 @@ boolean compilationLimitsEnabled() {
 
     void registerClusterSettingsListeners(ClusterSettings clusterSettings) {
         clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_SIZE_IN_BYTES, this::setMaxSizeInBytes);
-        clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_COMPILATIONS_RATE, compiler::setMaxCompilationRate);
+        clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED,
+                // Don't deref potentially null generalCache
+                r -> {
+                    if (generalCache != null) {
+                        generalCache.setMaxCompilationRate(r);
+                    }
+                },
+                s -> {
+                    if (generalCache == null) {
+                        // This validator is run on every settings update.
+                        // Here we try to detect if it's a spurious update to the default.
+                        // This _will_ allow through bad updates to the default.
+                        if (SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED.getDefault(Settings.EMPTY).equals(s) &&
+                            SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED.exists(settings) == false) {
```

Comment:
After looking through the code more carefully, we can't do this because settings is never actually updated for dynamic settings. This would only determine if this setting exists at start-up. So if a user were to dynamically update this setting we would be using old information.

Reply:
What do you think about changing this to:

`SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED.getDefault(Settings.EMPTY).equals(s)`

That will detect non-changes.

Reply:
That would only detect if the value was different than the default. The user could explicitly set the default through a dynamic update?

Reply:
Yes, the user can set that and we will not reject it.  I address this in the comment on [line 397](https://github.com/elastic/elasticsearch/pull/52855/files#diff-e7f789c9bbf345278cfb259a0b9ac87fR397):
```
// This validator is run on every settings update.
// Here we try to detect if it's a spurious update to the default.
// This _will_ allow through bad updates to the default.
```

Reply:
@stu-elastic Thanks for walking me through this. I believe you are correct that if we remove the exists check, this will behave as expected.

Reply:
Now we have a special `use-context` value.